src/logic.rs

```diff
@@ -8,141 +6,112 @@ use cardano_serialization_lib::metadata::{
-    // Group the registrations by stake key (each stake key may have one valid registration)
-    let mut m = HashMap::new();
-    for rego in regos {
-        let stake_key = rego.metadata.stake_vkey.clone();
-        m.entry(stake_key).or_insert_with(Vec::new).push(rego);
-    }
-    // Find the regos with the highest slot number, and of those, choose the
-    // lowest txid.
-    let mut latest_regos = Vec::new();
-    for (_, stake_regos) in m {
-        let latest = stake_regos
-            .iter()
-            .fold(stake_regos[0].clone(), |acc, rego| {
-                use std::cmp::Ordering::*;
-                match &rego.metadata.slot.cmp(&acc.metadata.slot) {
-                    // If the slot number is less, it's not a newer registration.
-                    Less => acc,
-                    // If the slot number is greater, it's a newer registration.
-                    Greater => rego.clone(),
-                    // If the slot number is equal, choose the one with the lower tx id.
-                    Equal => {
-                        if rego.tx_id < acc.tx_id {
-                            rego.clone()
-                        } else {
-                            acc
-                        }
-                    }
-                }
-            });
-        latest_regos.push(latest.clone())
-    }
-    latest_regos
+    Ok(reg_voting_power)
 }
 
 #[instrument]
 pub(crate) fn get_stake_address(
     stake_vkey_hex: &StakeVKey,
-    network_info: &NetworkInfo,
+    network: &NetworkInfo,
 ) -> Result<String> {
-    // Remove initial '0x' from string
-    if !stake_vkey_hex.starts_with("0x") {
-        warn!("stake_vkey_hex doesn't start with `0x`");
-    }
-
-    let stake_vkey_hex_only = stake_vkey_hex.trim_start_matches("0x");
+    let stake_vkey_hex = stake_vkey_hex.trim_start_matches("0x");
     // TODO support stake extended keys
-    if stake_vkey_hex_only.len() == 128 {
+    if stake_vkey_hex.len() == 128 {
         // TODO: why is this bad? can we give a better error here?
         bail!("stake_vkey has length 128");
     } else {
         // Convert hex to public key
-        let hex = hex::decode(&stake_vkey_hex_only)?;
+        let hex = hex::decode(&stake_vkey_hex)?;
         let pub_key = PublicKey::from_bytes(&hex).map_err(|_| eyre!(""))?;
         let cred = StakeCredential::from_keyhash(&pub_key.hash());
-        let stake_addr: Address = RewardAddress::new(network_info.network_id(), &cred).to_address();
+        let stake_addr = RewardAddress::new(network.network_id(), &cred).to_address();
         let stake_addr_bytes = stake_addr.to_bytes();
         let stake_addr_bytes_hex = hex::encode(&stake_addr_bytes);
         Ok(stake_addr_bytes_hex)
     }
 }
 
-impl Rego {
+impl Reg {
     /// Checks if this registration is valid
     ///
     /// Returns `Result` rather than `bool` to allow structured failures
     #[instrument]
```

Comment:
Out of curiosity, is the `bail!(*)` output consumed by the global `tokio` collector. I cannot spot the regular tokio logging levels here.

Reply:
`bail!(tokens)` actually just expands to: `return Err(eyre!(tokens))`, it's not really doing anything fancy, not sure what you mean about tokio logging levels. 😕

Reply:
sorry, I was referring to [instrument ](https://github.com/input-output-hk/voting-tools-rs/blob/3c3f3760dad0f8185a4bb9abd6099a1dac198daf/src/logic.rs#L98) and wondering if `tokio-tracing` picks up and collects the output of bail. I can now see that bail just bubbles the error up rather than doing logging et al.

Reply:
@cong-or the `instrument` attribute doesn't do any logging by itself, to actually pick up any data coming from that, a `Subscriber` is needed. We don't have a fully featured tracing implementation yet, but we're working on it.